timely/src/progress/subgraph.rs

```diff
@@ -530,26 +530,37 @@ where
         assert_eq!(self.children[0].outputs, self.inputs());
         assert_eq!(self.children[0].inputs, self.outputs());
 
-        let mut internal_summary = Vec::with_capacity(self.inputs());
-        for summary in self.scope_summary.iter() {
-            let scope_summary = summary.iter()
-                .map(|output| {
-                    output.elements()
-                        .iter()
-                        .cloned()
-                        .fold(
-                            Antichain::with_capacity(output.elements().len()),
-                            |mut antichain, path_summary| {
-                                antichain.insert(TInner::summarize(path_summary));
-                                antichain
-                            },
-                        )
-                })
-                .collect();
-
-            internal_summary.push(scope_summary);
+        // Note that we need to have `self.inputs()` elements in the summary
+        // with each element containing `self.outputs()` antichains regardless
+        // of how long `self.scope_summary` is
+        let mut internal_summary: Vec<Vec<_>> = (0..self.inputs())
+            .map(|_| (0..self.outputs()).map(|_| Antichain::new()).collect())
+            .collect();
+
+        for (summary, internal_summary) in
+            self.scope_summary.iter().zip(internal_summary.iter_mut())
+        {
+            internal_summary.reserve(summary.len());
+
+            for (output, antichain) in summary.iter().zip(internal_summary) {
+                antichain.reserve(output.elements().len());
+
+                for path_summary in output.elements() {
+                    antichain.insert(TInner::summarize(path_summary.clone()));
+                }
+            }
         }
```

Comment:
This might be because of the recent confusion, but at the moment I def prefer something more explicit like what was here before
```
        for input in 0 .. self.scope_summary.len() {
            for output in 0 .. self.scope_summary[input].len() {
                for path_summary in self.scope_summary[input][output].elements().iter() {
                    internal_summary[input][output].insert(TInner::summarize(path_summary.clone()));
                }
            }
        }
```
vs the zipping of iterators that might terminate early if they don't have the same lengths, etc.

Reply:
Well, the thing is that iteration (without indices) is faster than indexing because of the whole bounds checking thing, checks are never emitted with iteration/zipping but with manual indexing sometimes llvm drops the ball, it's your call though

Reply:
Let's go with clarity for the moment, and if we end up seeing CPU being a problem here ever, we could revisit it.